pkg/sources/s3/s3.go

```diff
@@ -132,79 +134,212 @@ func (s *Source) newClient(region string) (*s3.S3, error) {
-				bucketsToScan = append(bucketsToScan, *bucket.Name)
-			}
-		} else {
-			bucketsToScan = s.conn.Buckets
-		}
+	switch cred := s.conn.GetCredential().(type) {
+	case *sourcespb.S3_SessionToken:
+		baseCredentials = credentials.NewStaticCredentials(cred.SessionToken.Key, cred.SessionToken.Secret, cred.SessionToken.SessionToken)
+	case *sourcespb.S3_AccessKey:
+		baseCredentials = credentials.NewStaticCredentials(cred.AccessKey.Key, cred.AccessKey.Secret, "")
 	case *sourcespb.S3_Unauthenticated:
-		bucketsToScan = s.conn.Buckets
+		baseCredentials = credentials.AnonymousCredentials
+	case *sourcespb.S3_CloudEnvironment:
+		// Nothing needs to be done!
 	default:
-		return errors.Errorf("invalid configuration given for %s source", s.name)
+		return nil, errors.Errorf("invalid configuration given for %s source", s.name)
 	}
 
-	objectCount := uint64(0)
-	for i, bucket := range bucketsToScan {
-		if common.IsDone(ctx) {
-			return nil
-		}
 
-		s.SetProgressComplete(i, len(bucketsToScan), fmt.Sprintf("Bucket: %s", bucket), "")
+	sess, err := session.NewSession(cfg)
+	if err != nil {
+		return nil, err
+	}
+	
+	stsClient := sts.New(sess)
+	baseCredentials = stscreds.NewCredentialsWithClient(stsClient, roleArn, func(p *stscreds.AssumeRoleProvider) {
+		p.RoleSessionName = "trufflehog"
+	})
 
-		s.log.Info("Scanning bucket", "bucket", bucket)
-		region, err := s3manager.GetBucketRegionWithClient(context.Background(), client, bucket)
-		if err != nil {
-			s.log.Error(err, "could not get s3 region for bucket", "bucket", bucket)
-			continue
-		}
-		var regionalClient *s3.S3
-		if region != defaultAWSRegion {
-			regionalClient, err = s.newClient(region)
-			if err != nil {
-				s.log.Error(err, "could not make regional s3 client")
-			}
-		} else {
-			regionalClient = client
-		}
+	cfg.Credentials = baseCredentials
 
-		errorCount := sync.Map{}
+	sess, err = session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+		Config: *cfg,
+	})
 
-		err = regionalClient.ListObjectsV2PagesWithContext(
-			ctx, &s3.ListObjectsV2Input{Bucket: &bucket},
-			func(page *s3.ListObjectsV2Output, last bool) bool {
-				s.pageChunker(ctx, regionalClient, chunksChan, bucket, page, &errorCount, i+1, &objectCount)
-				return true
-			})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create and return the S3 client with the final configuration
+	return s3.New(sess), nil
+}
 
-		if err != nil {
-			return fmt.Errorf(
-				"could not list objects in s3 bucket: bucket %s: %w",
-				bucket,
-				err)
+// Chunks emits chunks of bytes over a channel.
+func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) error {
+	const defaultAWSRegion = "us-east-1"
+
+	if len(s.conn.Roles) > 0 {
+		for _, role := range s.conn.Roles {
+			
+		        client, err := s.newRoleClient(defaultAWSRegion, role)
+		        if err != nil {
+		        	return errors.WrapPrefix(err, "could not create s3 client", 0)
+		        }
+
```

Comment:
**Suggestion:** Moving the code in the loop body could help with readability, especially as the steps grow. It also helps to see what is needed for the operation and makes the code generally more testable.

```go
for _, role := range s.conn.Roles {
    // adjust name as desired
    if err := s.chunkWithRole(ctx, chunksChan, defaultAWSRegion, role); err != nil {
        ctx.Logger().V(2).Info("failed chunking with role", "role", role)
        continue
    }
}
```